autogen/numpy_api_dump.py

```diff
@@ -277,9 +256,6 @@ def find_common_type(array_types, scalar_types):
 def fix(x, out=None):
     raise NotImplementedError
 
```

Comment:
Even more, you may want to consider creating a type system and using it here. In PrimTorch it was really helpful because of the linter and to catch errors early, but also when doing metaprogramming.

Reply:
For this file specifically, I'm not sure it's worth linting: it's an autogenerated dump of the numpy namespace, made only to simplify clicking around the numpy API docs. I can take this off the repository if it gets in the way.

Reply:
With type hints, I'm curious on specifics of how they helped as you say, and at which stage of the lifetime of the project. I've not used them yet, and had an impression that they are more hassle than worth.

Reply:
Black formatting is pretty cheap and better to do earlier in the lifetime of the project. Typing I am a bit more ambivalent about; it works well if you are willing to put in the effort and have complete control over API design, but a lot of Pythonic API style isn't easily typeable and you have to be willing to not do it, or put up with suppressions

Reply:
About how type hints helped: it was helpful to express the semantics of the functions that we had. It made the code much easier to read and reason about. It also exposed bugs early. I mean, like any other type system really.

It's also true that it takes a bit of effort to maintain, as Ed mentioned. 

In any case, I'm +1 on black and +0 on the type system.

Reply:
+1 for adding `black` too, at some point before merging code where it needs to be maintained long term.

> it works well if you are willing to put in the effort and have complete control over API design, but a lot of Pythonic API style isn't easily typeable and you have to be willing to not do it, or put up with suppressions

Indeed. When starting from scratch, type annotations are great - exactly because they help you avoid problematic constructs (the kinds that aren't easy to statically type, and those NumPy has a ton of). For this project we have to adhere to the NumPy API, and we can't really avoid array-like inputs, dtype keywords that take multiple kinds of types, polymorphic functions like `unique` and `cross`, etc. 

So here, I'd be neutral at best on type annotations. At least as far as exposing them goes - if they're internal only and some shortcuts can be taken when the above-mentioned constructs are encountered, then they may still be helpful.

Reply:
One maintenancy consideration of blackening (is that a word?) is gh-10. tests vendored from numpy should likely by excluded to minimize the diff size going forward.